blackbox/handlers/storage/google_drive.py

```diff
@@ -0,0 +1,252 @@
+
+        Important note: If a folder with this name already exists, another will be
+        created. The original will not be overwritten.
+
+        Args
+            folder_path: The path to the folder. It does not matter if there is a
+                         leading or trailing slash.
+            parent_id: The ID of the parent folder. Default is "root".
+
+        Return
+            The ID of the created folder.
+        """
+
+        folder_names = folder_path.split("/")
+        for folder_name in folder_names:
+            metadata = {
+                "name": folder_name,
+                "mimeType": "application/vnd.google-apps.folder",
+                "parents": [parent_id],
+            }
+            folder = self.client.files().create(
+                body=metadata,
+                fields="id",
+            ).execute()
+            parent_id = folder.get("id")
+        return parent_id
+    
+    def _get_and_ensure_deepest_folder_id(self, path: str) -> str:
+        """
+        Get the ID of the deepest folder in the provided path.
+
+        Any folders in the path that do not already exist will be created.
+
+        Args
+            path: The folder path, excluding a file. It does not matter if there is a
+                  leading or trailing slash.
+        
+        Return
+            The ID of the deepest folder.
+        """
+
+        deepest_folder = path.split("/")[-1]  # Get the last part of the path
+        folder = self._find_folder(path)  # Get the folder object from Google Drive
+        if not folder or folder["name"] != deepest_folder:
+            # Create all folders in the path if none exist. Otherwise, only create
+            # the ones that don't exist yet.
+            new_folder_path = path if not folder else path.split(folder["name"])[1]
+            new_folder_path = (new_folder_path[1:]  # Remove leading slash
+                               if new_folder_path.startswith("/")
+                               else new_folder_path)
+            parent_id = "root" if not folder else folder["id"]
+            return self._create_folder(
+                folder_path=new_folder_path,
+                parent_id=parent_id,
+            )
+        else:
+            return folder["id"]
+
+    def _upload(self, file_path: str, file_content: bytes | str) -> str:
+        """
+        Upload a file to Google Drive.
+
+        Args
+            file_path: The path to upload the file to.
+            file_content: The content of the file to upload.
+    
+        Return
+            The ID of the uploaded file.
+        """
+
+        # Determine the MIME type of the file, because we need to include this in the
+        # payload when we upload the file to Google Drive.
+        mimetype, _ = mimetypes.guess_type(file_path)
+
+        # Get the folder
+        folder_id = "root"  # Use the root folder as a default
+        folder_path = "/".join(file_path.split("/")[:-1])  # Path excluding filename
```

Comment:
this again seems a bit dubious if there's a trailing slash, it would just return `' '`.